clorm/orm/core.py

```diff
@@ -2360,115 +2357,223 @@ def __iter__(self):
-    except (TypeError,AttributeError,RuntimeError):
-        raise ValueError(("Failed to unify clingo.Symbol object {} with "
-                          "Predicate class {}").format(raw, cls.__name__))
-
-# Construct a Predicate via the field keywords
-def _predicate_init_by_keyword_values(self, **kwargs):
-    argnum=0
-    field_values = []
-    clingoargs = []
-    for f in self.meta:
-        if f.name in kwargs:
-            v= _preprocess_field_value(f.defn, kwargs[f.name])
-            argnum += 1
-        elif f.defn.has_default:
-            # Note: must be careful to get the default value only once in case
-            # it is a function with side-effects.
-            v = _preprocess_field_value(f.defn, f.defn.default)
+        raw_args = raw.arguments
+        if len(raw_args) != {pdefn.arity}:
+            return None
+
+        {{%sign_check_unify%}}
+
+        if raw.name != "{pdefn.name}":
+            return None
+
+        return cls({{%args_cltopy%}}
+                   raw=raw, sign=raw.positive)
+    except (TypeError, ValueError):
+        return None
+    except AttributeError as e:
+        raise ValueError((f"Cannot unify with object {{raw}} ({{type(raw)}}) as "
+                          "it is not a clingo Symbol Function object"))
+
+"""
+
+CHECK_SIGN_TEMPLATE = r"""
+# Check if the sign is allowed
+if self._sign != {sign}:
+    raise ValueError(f"Predicate {{type(self).__name__}}"
+                     f"is defined to only allow {pdefn.sign} instances")
+"""
+
+CHECK_SIGN_UNIFY_TEMPLATE = r"""
+if raw.positive != {sign}:
+    return None
+"""
+
+NO_DEFAULTS_TEMPLATE = r"""
+# Check for missing values that have no defaults
+if MISSING in ({args}):
+    for arg, name in ({named_args}):
+        if arg == MISSING:
+            raise TypeError((f"Missing argument for field \"{{name}}\""
+                             f"(which has no default value)"))
+"""
+
+ASSIGN_DEFAULT_TEMPLATE = r"""
+if {arg} == MISSING:
+    {arg} = {arg}_field.default
+"""
+
+ASSIGN_COMPLEX_TEMPLATE = r"""
+if not isinstance({arg}, {arg}_class):
+    {arg} = _instance_from_tuple({arg}_class, {arg})
+"""
+
+PREDICATE_UNIFY_DOCSTRING = r"""
+    Unify a (raw) clingo.Symbol object with the class.
+
+    Returns None on failure to unify otherwise returns the new fact
+"""
+
+def _generate_dynamic_predicate_functions(class_name: str, namespace: Dict):
+    pdefn = namespace["_meta"]
+
+    gdict = {"_instance_from_tuple": _instance_from_tuple,
+             "_get_symbols": _get_symbols,
+             "MISSING": MISSING,
+             "AnySymbol": AnySymbol,
+             "Predicate": Predicate}
+
+    for f in pdefn:
+        gdict[f"{f.name}_field"] = f.defn
+        gdict[f"{f.name}_pytocl"] = f.defn.pytocl
+        gdict[f"{f.name}_cltopy"] = f.defn.cltopy
+
+        if f.defn.complex:
+            gdict[f"{f.name}_class"] = f.defn.complex
+
+    args_signature = "".join([f"{f.name}=MISSING, " for f in pdefn])
```

Comment:
I really like that you introduced the use of MISSING. It seems perfect to use here as it provides the flexibility to have any combination of default and non-default arguments and the position doesn't matter. 

But adding the actual (non-callable) defaults is a good idea. Although not sure I have the energy to do it today :)

Reply:
I was just thinking that maybe it could be done after all, but a bit different. For each (non-callable) default, you add an entry for the instance in the global dict with some made up name. So you would end up with something a string like `a=a_default, b=b_default` and the global dict would have an entry `gdict[f"{f.name}_default"] = f.defn.default`. This way you never have to turn the instance into a string representation and have to parse in back into instantiating the object. Maybe this is what you were already thing.

Reply:
I think you are one of the developers of dataclasses, they are doing the same thing there^^